src/Env.hs

```diff
@@ -1,53 +1,61 @@
-{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveFunctor      #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
-{-# LANGUAGE LambdaCase #-}
-{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE DerivingStrategies #-}
+
 module Env (Env, empty, insert, singleton, lookup, lookupIdent, lookupVal, toList, fromList, named) where
 
 import Prelude hiding (lookup)
 
 import Control.Lens
-import Data.Map (Map)
-import qualified Data.Map as Map
+import Data.IntMap.Strict (IntMap)
+import qualified Data.IntMap.Strict as IM
 import Data.Text (Text)
 
 import Syntax (Ident, Stx(..))
 
-newtype Env v a = Env (Map v (Ident, a))
-  deriving (Eq, Functor, Monoid, Semigroup, Show)
+import Util.Key
+
+-- | an environment is a mapping to a map of @Ident@s and some value. We keep
+-- 'v' as a phantom to stay honest
```

Comment:
Neither the comment nor the definition explains what `v` is supposed to represent. How about something like this:

```suggestion
-- | An environment maps variables of type 'v' to values of type 'a'.
--
-- More specifically, 'Env' is intended to be used for variables in the core
-- language, such as 'Var' and 'MacroVar', which are represented as unique
-- integers in order to avoid accidental capture. To make error message more
-- readable, 'Env' also stores the original 'Ident' in addition to the value.
```